components/node-labeler/cmd/run_test.go

```diff
@@ -0,0 +1,254 @@
+// Copyright (c) 2025 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License.AGPL.txt in the project root for license information.
+
+package cmd
+
+import (
+	"context"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/aws/smithy-go/ptr"
+	csapi "github.com/gitpod-io/gitpod/content-service/api"
+	workspacev1 "github.com/gitpod-io/gitpod/ws-manager/api/crd/v1"
+	"github.com/golang/mock/gomock"
+	"github.com/google/uuid"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"google.golang.org/protobuf/proto"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/envtest"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const (
+	timeout            = time.Second * 10
+	duration           = time.Second * 2
+	interval           = time.Millisecond * 250
+	workspaceNamespace = "default"
+	secretsNamespace   = "workspace-secrets"
+)
+
+var (
+	k8sClient     client.Client
+	testEnv       *envtest.Environment
+	mock_ctrl     *gomock.Controller
+	ctx           context.Context
+	cancel        context.CancelFunc
+	workspaceCtrl *WorkspaceCountController
+	NodeName      = "cool-ws-node"
+)
+
+func TestAPIs(t *testing.T) {
+	mock_ctrl = gomock.NewController(t)
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "Controller Suite")
+}
+
+var _ = Describe("WorkspaceCountController", func() {
```

Comment:
We have this disappearing node test for mk2.

https://github.com/gitpod-io/gitpod/blob/bd36bac035c22a7b519f3e895c040aaa17916fa0/components/ws-manager-mk2/controllers/workspace_controller_test.go#L296

It would be interesting to see if a similar test could be added for node-labeler, that way we could do a test like for this: https://github.com/gitpod-io/gitpod/pull/20509#issuecomment-2588007398

Reply:
The mk2 test you linked here is using the k8s client to delete the node:
https://github.com/gitpod-io/gitpod/blob/bd36bac035c22a7b519f3e895c040aaa17916fa0/components/ws-manager-mk2/controllers/workspace_controller_test.go#L346-L348

I have two questions here:
1. Is that analogous to a node being ripped to shreds from the cloud provider's side?
2. I'm probably missing something, but if the node's gone, what is there to annotate? Do we need to care about a node disappearing at all?

Reply:
I'm not sure. I think we need to do a brief manual test, where we delete a node from the cloud provider while it has a workspace on it, and see how k8s reconciles the delete node with the scale down disabled: true annotation